File: ios/Classes/AgoraRtcVoiceEnginePlugin.swift
import Flutter
import UIKit
import AgoraRtcKit

public class AgoraRtcVoiceEnginePlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  var eventSink: FlutterEventSink?
      var agoraEngine: AgoraRtcEngineKit?

      public static func register(with registrar: FlutterPluginRegistrar) {
          let methodChannel = FlutterMethodChannel(name: "com.danieluwadi.agora_rtc_voice_engine/methods", binaryMessenger: registrar.messenger())
          let eventChannel = FlutterEventChannel(name: "com.danieluwadi.agora_rtc_voice_engine/events", binaryMessenger: registrar.messenger())

          let instance = AgoraRtcVoiceEnginePlugin()
          registrar.addMethodCallDelegate(instance, channel: methodChannel)
          eventChannel.setStreamHandler(instance)
      }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)

    case "initialize":
      guard let args = call.arguments as? [String: Any],
            let appId = args["appId"] as? String else {
          result(FlutterError(code: "INVALID_ARGS", message: "App ID is required", details: nil))
          return
      }
      let areaCode = args["areaCode"] as? String
      initializeAgora(appId: appId, areaCode: areaCode)
      result(nil)

    case "release":
        agoraEngine?.leaveChannel(nil)
        agoraEngine = nil
        result(nil)

    case "joinChannel":
        let args = call.arguments as! [String: Any]
        let token = args["token"] as? String
        let channelName = args["channelName"] as! String
        let uid = args["uid"] as? UInt ?? 0
        agoraEngine?.joinChannel(byToken: token, channelId: channelName, info: nil, uid: uid, joinSuccess: nil)
        result(nil)

    case "leaveChannel":
        agoraEngine?.leaveChannel(nil)
        result(nil)

    case "enableAudioVolumeIndication":
        let args = call.arguments as! [String: Any]
        let interval = args["interval"] as? Int ?? 200
        let smooth = args["smooth"] as? Int ?? 3
        let reportVad = args["reportVad"] as? Bool ?? false
        agoraEngine?.enableAudioVolumeIndication(interval, smooth: smooth, report_vad: reportVad)
        result(nil)

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func initializeAgora(appId: String, areaCode: String?) {
          agoraEngine = AgoraRtcEngineKit.sharedEngine(withAppId: appId, delegate: self)
          // areaCode isn't directly applied in iOS SDK init, handled via channel profile/config if needed
  }

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
          self.eventSink = events
          return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
      self.eventSink = nil
      return nil
  }
}

extension AgoraRtcVoiceEnginePlugin: AgoraRtcEngineDelegate {
    public func rtcEngine(_ engine: AgoraRtcEngineKit, didJoinedOfUid uid: UInt, elapsed: Int) {
        eventSink?(["event": "onUserJoined", "uid": uid])
    }

    public func rtcEngine(_ engine: AgoraRtcEngineKit, didOfflineOfUid uid: UInt, reason: AgoraUserOfflineReason) {
        eventSink?(["event": "onUserOffline", "uid": uid])
    }

    public func rtcEngine(_ engine: AgoraRtcEngineKit, reportAudioVolumeIndicationOfSpeakers speakers: [AgoraRtcAudioVolumeInfo], totalVolume: Int) {
        let list = speakers.map { ["uid": $0.uid, "volume": $0.volume] }
        eventSink?(["event": "onAudioVolumeIndication", "speakers": list])
    }
}

File: android/src/main/kotlin/com/danieluwadi/agora_rtc_voice_engine/AgoraRtcVoiceEnginePlugin.kt
package com.danieluwadi.agora_rtc_voice_engine

import android.content.Context
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.agora.rtc2.*

class AgoraRtcVoiceEnginePlugin: FlutterPlugin, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {
  private lateinit var methodChannel: MethodChannel
  private lateinit var eventChannel: EventChannel
  private var eventSink: EventChannel.EventSink? = null
  private var rtcEngine: RtcEngine? = null
  private lateinit var context: Context

  override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    context = binding.applicationContext
    methodChannel = MethodChannel(binding.binaryMessenger, "com.danieluwadi.agora_rtc_voice_engine/methods")
    methodChannel.setMethodCallHandler(this)

    eventChannel = EventChannel(binding.binaryMessenger, "com.danieluwadi.agora_rtc_voice_engine/events")
    eventChannel.setStreamHandler(this)
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    methodChannel.setMethodCallHandler(null)
    eventChannel.setStreamHandler(null)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    when (call.method) {
      "initialize" -> {
        val appId = call.argument<String>("appId")!!
        rtcEngine = RtcEngine.create(context, appId, object : IRtcEngineEventHandler() {
          override fun onUserJoined(uid: Int, elapsed: Int) {
            eventSink?.success(mapOf("event" to "onUserJoined", "uid" to uid))
          }

          override fun onUserOffline(uid: Int, reason: Int) {
            eventSink?.success(mapOf("event" to "onUserOffline", "uid" to uid))
          }

          override fun onAudioVolumeIndication(speakers: Array<out AudioVolumeInfo>?, totalVolume: Int) {
            val list = speakers?.map { mapOf("uid" to it.uid, "volume" to it.volume) } ?: emptyList()
            eventSink?.success(mapOf("event" to "onAudioVolumeIndication", "speakers" to list))
          }
        })
        rtcEngine?.setChannelProfile(Constants.CHANNEL_PROFILE_COMMUNICATION)
        result.success(null)
      }
      "joinChannel" -> {
        val token = call.argument<String>("token")!!
        val channelName = call.argument<String>("channelName")!!
        val uid = call.argument<Int>("uid") ?: 0
        rtcEngine?.joinChannel(token, channelName, null, uid)
        result.success(null)
      }
      "leaveChannel" -> {
        rtcEngine?.leaveChannel()
        result.success(null)
      }
      "getPlatformVersion" -> {
        result.success("Android ${android.os.Build.VERSION.RELEASE}")
      }
      else -> result.notImplemented()
    }
  }

  override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
    eventSink = events
  }

  override fun onCancel(arguments: Any?) {
    eventSink = null
  }
}